pkg/controller/history/controller_history.go

```diff
@@ -0,0 +1,487 @@
+		ref := controller.GetControllerOf(history[i])
+		if ref == nil || ref.UID == parent.GetUID() {
+			owned = append(owned, history[i])
+		}
+
+	}
+	return owned, err
+}
+
+func (rh *realHistory) CreateControllerRevision(parent metav1.Object, revision *apps.ControllerRevision) (*apps.ControllerRevision, error) {
+	// Initialize the probe to 0
+	probe := uint32(0)
+
+	// Clone the input
+	any, err := scheme.Scheme.DeepCopy(revision)
+	if err != nil {
+		return nil, err
+	}
+	clone := any.(*apps.ControllerRevision)
+
+	// Continue to attempt to create the revision updating the name with a new hash on each iteration
+	for {
+		var hash uint32
+		// The first attempt uses no probe to resolve collisions
+		if probe > 0 {
+			hash = HashControllerRevision(revision, &probe)
+		} else {
+			hash = HashControllerRevision(revision, nil)
+		}
+		// Update the revisions name and labels
+		clone.Name = ControllerRevisionName(parent.GetName(), hash)
+		created, err := rh.client.Apps().ControllerRevisions(parent.GetNamespace()).Create(clone)
+		if errors.IsAlreadyExists(err) {
+			probe++
+			continue
+		}
+		return created, err
+	}
+}
+
+func (rh *realHistory) UpdateControllerRevision(revision *apps.ControllerRevision, newRevision int64) (*apps.ControllerRevision, error) {
+	obj, err := scheme.Scheme.DeepCopy(revision)
+	if err != nil {
+		return nil, err
+	}
+	clone := obj.(*apps.ControllerRevision)
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if clone.Revision == newRevision {
+			return nil
+		}
+		clone.Revision = newRevision
+		_, updateErr := rh.client.Apps().ControllerRevisions(clone.Namespace).Update(clone)
+		if updateErr == nil {
+			return nil
+		}
+		if updated, err := rh.lister.ControllerRevisions(clone.Namespace).Get(clone.Name); err == nil {
+			// make a copy so we don't mutate the shared cache
+			obj, err := scheme.Scheme.DeepCopy(updated)
+			if err != nil {
+				return err
+			}
+			clone = obj.(*apps.ControllerRevision)
+		}
+		return updateErr
+	})
+	return clone, err
+}
+
+func (rh *realHistory) DeleteControllerRevision(revision *apps.ControllerRevision) error {
+	return rh.client.Apps().ControllerRevisions(revision.Namespace).Delete(revision.Name, nil)
+}
+
+func (rh *realHistory) AdoptControllerRevision(parent metav1.Object, parentKind schema.GroupVersionKind, revision *apps.ControllerRevision) (*apps.ControllerRevision, error) {
+	// Return an error if the parent does not own the revision
+	if owner := controller.GetControllerOf(revision); owner != nil {
+		return nil, fmt.Errorf("attempt to adopt revision owned by %v", owner)
+	}
+	// Use strategic merge patch to add an owner reference indicating a controller ref
+	return rh.client.Apps().ControllerRevisions(parent.GetNamespace()).Patch(revision.GetName(),
+		types.StrategicMergePatchType, []byte(fmt.Sprintf(
+			`{"metadata":{"ownerReferences":[{"apiVersion":"%s","kind":"%s","name":"%s","uid":"%s","controller":true,"blockOwnerDeletion":true}],"uid":"%s"}}`,
```

Comment:
Yeah, this is needed in ReplicaSets so the deletion can cascade to Pods but not in ControllerRevisions (at least for now) - disregard.